server/src/main/java/io/crate/expression/scalar/string/ParseURLFunction.java

```diff
@@ -0,0 +1,142 @@
+import io.crate.types.DataTypes;
+
+public final class ParseURLFunction extends Scalar<Object, String> {
+
+    private static final String NAME = "parse_url";
+
+    public static void register(ScalarFunctionModule module) {
+        module.register(
+            Signature.scalar(
+                NAME,
+                DataTypes.STRING.getTypeSignature(),
+                DataTypes.UNTYPED_OBJECT.getTypeSignature()
+            ),
+            ParseURLFunction::new
+        );
+    }
+
+    private final Signature signature;
+    private final BoundSignature boundSignature;
+
+    public ParseURLFunction(Signature signature, BoundSignature boundSignature) {
+        this.signature = signature;
+        this.boundSignature = boundSignature;
+    }
+
+    @Override
+    public Signature signature() {
+        return signature;
+    }
+
+    @Override
+    public BoundSignature boundSignature() {
+        return boundSignature;
+    }
+
+    @Override
+    @SafeVarargs
+    public final Object evaluate(TransactionContext txnCtx, NodeContext nodeCtx, Input<String>... args) {
+        String url = args[0].value();
+        if (url == null) {
+            return null;
+        }
+        return parseURL(url);
+    }
+
+    private final Object parseURL(String urlText) {
+        final Map<String, Object> urlMap = new LinkedHashMap<>();
+
+        URL url = null;
+
+        try {
+            url = new URL(urlText);
+        } catch (MalformedURLException e1) {
+            throw new IllegalArgumentException(String.format(Locale.ENGLISH,
+                                                            "unable to parse url %s",
+                                                             urlText));
+        }
+
+        urlMap.put("scheme", url.getProtocol());
+        urlMap.put("userinfo", decodeText(url.getUserInfo()));
+        urlMap.put("hostname", url.getHost());
+        urlMap.put("port", url.getPort() == -1 ? null : url.getPort());
+        urlMap.put("path", decodeText(url.getPath()));
+        urlMap.put("query", decodeText(url.getQuery()));
+        urlMap.put("parameters", parseQuery(url.getQuery()));
+        urlMap.put("fragment", decodeText(url.getRef()));
+
+        return urlMap;
+    }
+
+    private final Map<String,List<String>> parseQuery(String query) {
+        if (Strings.isNullOrEmpty(query)) {
+            return null;
+        }
+
+        Map<String,List<String>> queryMap = new HashMap<String,List<String>>();
+        Arrays.stream(query.split("&(?!amp)"))
+                     .forEach((String param) -> parseQueryParameters(queryMap, param));
+
+        return queryMap;
+    }
+
+    private final void parseQueryParameters(Map<String,List<String>> map, String params) {
+        final int idx = params.indexOf("=");
+        final String key = idx > 0 ? decodeText(params.substring(0, idx)) : decodeText(params);
+        final String value = idx > 0 && params.length() > idx + 1 ? decodeText(params.substring(idx + 1)) : null;
```

Comment:
Not sure but I think values can be a comma separated list, ie, we can pass to `params` string looking like 
`key = val1, val2` which it still fine to treat as a single value even in object. 

We could mention explicitly in docs that comma separated list is NOT converted into array in the `parsed_url_obj[params]` and and a test case. 

Or we can convert comma separated lists to array if you think it's more convenient to users, no strong opinion here.

Reply:
Converting to an array is not easy, as it would make it impossible to use a specific type and index it, which is the main purpose of the function.

It is part of the documentation, that the result is an `ARRAY(TEXT)` for each dedicated parameter (key)

Basically one should be able to do the following:

```sql
SELECT * FROM tab WHERE 'hello' = ANY(url['parameters']['my_parameter'])
```

Reply:
>It is part of the documentation, that the result is an ARRAY(TEXT)

I saw 
```
"parameters" OBJECT AS (
             "key1" ARRAY(TEXT),
```
However, it might be still not clear from docs (and also not covered in tests) that _each item_ of the `ARRAY` can be a comma separated list. Ie for query string
`key1=single_val&key2=val2&key1=x,y,z` we end up having 
```
parameters {
    key1: ["single_val", "x,y,z"],
    key2:val2
}

```
and NOT
```
parameters {
    key1: ["single_val", "x", "y", "z"],
    key2:val2
}
```

Reply:
How about adding another example to the docs then?
```sql
cr> SELECT parse_url('http://crate.io?p1=1,2,3&p1=a&p1')['parameters'] as params;                                                                                                                                
+------------------------------+
| params                       |
+------------------------------+
| {"p1": ["1,2,3", "a", null]} |
+------------------------------+
SELECT 1 row in set (0.002 sec)
```

Reply:
Sounds good. Maybe add a note before this example, that each item of the array can be a comma separated list

Reply:
Added the example with note

```rst
Parameter values are always treated as ``text``. There is no conversion of 
comma-separated parameter values into arrays::

    cr> SELECT parse_url('http://crate.io?p1=1,2,3&p1=a&p2[]=1,2,3')['parameters'] as params;                                                                                                                        
    +-------------------------------------------+
    | params                                    |
    +-------------------------------------------+
    | {"p1": ["1,2,3", "a"], "p2[]": ["1,2,3"]} |
    +-------------------------------------------+
    SELECT 1 row in set (... sec)
```

---

I also checked, but there is no standard way to handle conversions anyway -> very framework specific.

Reply:
> However, it might be still not clear from docs (and also not covered in tests) that each item of the ARRAY can be a comma separated list.

According to the URL RFC/Specifications, a query is just defined as an arbitrary string. Even the use of attribute-value pairs is not strongly defined. I would not go such deep and explain well-known behaviour of query parameters in our documentation, thus adding an example about a valid value containing `,` chars is way too much for our docs in my opinion.

Reply:
I saw `"key1" ARRAY(TEXT)` in docs and supposed that on can treat it as lists are split into single items - because it's an implementation details and we could have done it in that way and I thought a small clarification is fine. Not insisting though